Step 3 - 데이터베이스 적용

kimjhyun0627 opened the pull request:
## ✨feat: 데이터베이스 적용

### 구현 기능: 데이터베이스 적용

- step 2 리뷰 반영해보았습니다(record class, setter 명시적 함수명 사용)!
- H2 데이터베이스를 연결했습니다
- JdbcTemplate와 SinpleJdbcIntert를 사용한 JdbcProductRepository를 구현했습니다

궁금증: Intellij를 사용하고 있는데, schema.sql을 제외한 모든 SQL 구분에
![image](https://github.com/user-attachments/assets/136a49d5-e4c0-4a1b-bfd0-76e686832e0d)
해당 오류가 발생합니다..!
프로젝트 구현에는 문제가 없었으나, 뭔가 인텔리제이의 모든 기능을 사용하지 못하고 있다는 생각이 들어
혹시 조언을 주실 수 있으면 정말 큰 도움이 될 것 같습니다 !

kimjhyun0627 committed:
docs: add feature list of step3 to README.md

kimjhyun0627 committed:
refactor: step 2 리뷰 반영

- record class로 변경: DTO, entity
- Product.java Setter 명시적으로 변경 -> with{OOO}

kimjhyun0627 committed:
feat: H2 데이터베이스 연결 설정

- H2 데이터베이스 연결 설정
- build 업데이트
- properties 업데이트

kimjhyun0627 committed:
feat: schema.sql 에 테이블 스키마 정의

- schema.sql 에 테이블 스키마 정의
- AUTO_INCREMENT 포함
- 불필요한 파일 및 코드 삭제

kimjhyun0627 committed:
feat: data.sql에 초기 데이터 삽입

- data.sql에 초기 데이터 삽입
- properties 업데이트

kimjhyun0627 committed:
feat: JdbcProductRepository 구현

- JdbcProductRepository.java 구현
- properties 수정
- 초기 data의 id값 삭제

liquidjoo commented on README.md:
```diff
@@ -21,4 +21,14 @@
\ No newline at end of file
+    * `DELETE /admin/products/{id}`
+
+## Step 3: 구현할 기능 목록
+
+* [X] step 2 리뷰 반영
+    * record class로 변경
+    * Product.java의 setter 함수명 명시적으로 변경
+* [X] H2 데이터베이스 연결 설정
+* [X] `schema.sql`에 테이블 스키마 정의 (AUTO_INCREMENT 포함)
+* [X] `data.sql`에 초기 데이터 삽입
+* [X] `JdbcTemplate` 기반 `ProductRepository` 구현
```
구현할 기능 목록 정의 좋습니다

liquidjoo commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -1,46 +1,14 @@
-    public String getImageUrl() {
-        return imageUrl;
-    }
+public record ProductRequest(
+        @NotBlank(message = "상품 이름은 필수 입력값입니다.") String name,
+        @Min(value = 1, message = "가격은 1원 이상이어야 합니다.") int price,
+        @NotBlank(message = "이미지 URL은 필수 입력값입니다.") String imageUrl
+) {
```
레코드 함수 활용 👍 
다양하게 접근하는 방법을 알고 있으면 개발에 많은 도움이 될 것 같아요

liquidjoo commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,54 +1,25 @@
-
-    public int getPrice() {
-        return price;
+public record Product(
+        Long id,
+        String name,
+        int price,
+        String imageUrl
+) {
+    
+    public Product withId(Long id) {
+        return new Product(id, this.name, this.price, this.imageUrl);
```
record 와 factory method 를 같이 활용하는 것에 대해서 고민해보면 좋을 것 같아요
`언제 record만 쓰고, 언제 factory method를 쓰면 좋은가` 에 대한 고민을 해보고 어떻게 처리하면 좋을지 고민해 보아요!

kimjhyun0627 replied:
factory method를 추가했습니다 !
id를 SQL에서 AUTO_INCREMENT로 생성하는 로직을 사용하기에, 생각보다는 해당 메소를 사용할 일이 많지는 않다는 것을 알게 되었습니다. 대신 새롭게 분리한 ProductRowMapper에서 id 로직 검사가 필요할 때 사용하고 있습니다 !

liquidjoo commented on src/main/java/gift/repository/JdbcProductRepository.java:
```diff
@@ -0,0 +1,80 @@
+                    rs.getString("name"),
+                    rs.getInt("price"),
+                    rs.getString("image_url")
+            );
```
RowMapper 를 별도로 관리하고 싶다면 별도의 객체로 만들어도 좋습니다 :)

liquidjoo replied:
별도로 나눈다면 어떤 장점이 있을까요? 한번 적어 보아요

kimjhyun0627 replied:
재사용성이 증가하고, 관심사 분리 역시 조금 더 깔끔하게 잘 될 것 같습니다! 명시적으로 ProductRowMapper라는 이름을 사용함으로써 명확하게 역할을 이해할 수도 있을 것 같습니다.

liquidjoo commented on src/main/java/gift/repository/JdbcProductRepository.java:
```diff
@@ -0,0 +1,80 @@
+            );
+            if (updated == 0) {
+                throw new IllegalArgumentException("상품을 찾을 수 없습니다: " + product.id());
+            }
```
단일책임원칙에 대해서 아시나요?
단일책임원칙(SRP)를 고민해보고 위의 코드를 어떻게 변경할 수 있을지 고민해 보면 좋을 것 같습니다

kimjhyun0627 replied:
단일 책임 원칙에 대해 학습한 내용을 토대로 코드 수정을 최소화하기 위해 insertProduct()와 updateProduct() 함수를 새로 만들어 save 함수에서 이를 불러오는 방식을 채택했습니다 !

liquidjoo commented on src/main/resources/db/init/data.sql:
```diff
@@ -0,0 +1,3 @@
+INSERT INTO PRODUCT (NAME, PRICE, IMAGE_URL)
+VALUES ('Sample1', 1000, 'http://example.com/1.png'),
+       ('Sample2', 2000, 'http://example.com/2.png');
```
빨간줄이 뜨는 상황이 어떠한 상황인지 자세히는 알수 없으나!
인텔리제이 셋팅으로 들어가서 Languages & Frameworks > SQL Dialects 를 h2 로 한번 설정해보시면 어떨까요?

kimjhyun0627 replied:
해결했습니다 !

liquidjoo reviewed (changes requested):
안녕하세요 진현님
3단계 구현 및 피드백 잘하셨어요 👍 
몇 가지 고민하면 좋을 것 같은 코멘트들 남겼습니다~
확인하고 다시 요청해 주세요 🙇

kimjhyun0627 committed:
refactor: step 3 - 코드 리뷰 반영

- Product.java: factory 메소드 추가
- ProductRowMapper.java 객체 분리: 추후 재사용성 및 명확성 증가
- JdbcProductRepository.java: save() SRP 적용: 하나의 함수가 너무 많은 책임을 지지 않도록 각각의 함수를 link해주는 형식으로 변경

kimjhyun0627 commented:
말씀해주신 코드 리뷰 바탕으로 커밋 진행하겠습니다 !

liquidjoo commented on src/main/java/gift/repository/JdbcProductRepository.java:
```diff
@@ -0,0 +1,85 @@
+    private final SimpleJdbcInsert insert;
+    private final RowMapper<Product> productRowMapper;
+
+    // 생성자를 통해 ProductRowMapper를 주입받습니다.
```
불필요한 주석은 제거해 주세요 🙇

liquidjoo commented on src/main/java/gift/repository/JdbcProductRepository.java:
```diff
@@ -0,0 +1,85 @@
+    public Optional<Product> findById(Long id) {
+        List<Product> list = jdbc.query(
+                "SELECT id, name, price, image_url FROM product WHERE id = ?", productRowMapper,
+                id); // 사용
```
불필요한 주석은 제거!

liquidjoo commented on src/main/java/gift/repository/ProductRowMapper.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ProductRowMapper implements RowMapper<Product> {
+
+    @Override
+    public Product mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return Product.createProduct(
+                rs.getLong("id"),
+                rs.getString("name"),
+                rs.getInt("price"),
+                rs.getString("image_url")
+        );
+    }
+}
```
별도의 역할로 분리한 점 좋습니다 👍

liquidjoo commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,54 +1,33 @@
-
-    public void setName(String name) {
-        this.name = name;
+    // Factory method
+    public static Product createProduct(Long id, String name, int price, String imageUrl) {
+        if (id == null) {
+            throw new IllegalArgumentException("Product ID cannot be null");
+        }
+        return new Product(id, name, price, imageUrl);
```
팩토리 메서드 패턴이 꼭 필요할까요? 고민해보면 좋을 것 같아요
정말 필요할 때는 언제인지!

liquidjoo reviewed (approved):
안녕하세요 진현님
피드백 반영 잘해주셨어요 👍 
조금 더 생각해보면 좋을 코멘트 하나 남겼습니다
코드는 잘 정리된 것 같아서 머지 진행할게요
저와 함께한 리뷰들이 도움이 되었으면 좋겠습니다
수고하셨습니다 🙇
